sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLPlannerImpl.java

```diff
@@ -130,30 +135,58 @@ public RelRoot rel(SqlNode sqlNode) throws RelConversionException {
   }
 
   public RelRoot rel(String sql, QueryParameters params) {
-    this.cluster = RelOptCluster.create(planner, new RexBuilder(typeFactory));
-    this.expressionConverter = new ExpressionConverter(cluster, params);
+    RelOptCluster cluster = RelOptCluster.create(planner, new RexBuilder(typeFactory));
 
     QueryTrait trait = new QueryTrait();
 
-    // Set up table providers that need to be pre-registered
-    // TODO(https://issues.apache.org/jira/browse/BEAM-8817): share this logic between dialects
-    List<List<String>> tables = Analyzer.extractTableNamesFromStatement(sql);
-    TableResolution.registerTables(this.defaultSchemaPlus, tables);
-
-    ResolvedStatement statement =
+    SqlAnalyzer analyzer =
         SqlAnalyzer.getBuilder()
             .withQueryParams(params)
             .withQueryTrait(trait)
             .withCalciteContext(config.getContext())
             .withTopLevelSchema(defaultSchemaPlus)
             .withTypeFactory((JavaTypeFactory) cluster.getTypeFactory())
-            .analyze(sql);
+            .build();
+
+    AnalyzerOptions options = SqlAnalyzer.initAnalyzerOptions(params);
+
+    // Set up table providers that need to be pre-registered
+    List<List<String>> tables = analyzer.extractTableNames(sql, options);
+    TableResolution.registerTables(this.defaultSchemaPlus, tables);
+    SimpleCatalog catalog =
+        analyzer.createPopulatedCatalog(defaultSchemaPlus.getName(), options, tables);
+
+    ImmutableMap.Builder<String, ResolvedCreateFunctionStmt> udfBuilder = ImmutableMap.builder();
+
+    ResolvedStatement statement;
+    ParseResumeLocation parseResumeLocation = new ParseResumeLocation(sql);
+    do {
+      statement = analyzer.analyzeNextStatement(parseResumeLocation, options, catalog);
+      if (statement.nodeKind() == RESOLVED_CREATE_FUNCTION_STMT) {
+        ResolvedCreateFunctionStmt createFunctionStmt = (ResolvedCreateFunctionStmt) statement;
+        // ResolvedCreateFunctionStmt does not include the full function name, so build it here.
+        String functionFullName =
+            String.format(
+                "%s:%s",
+                SqlAnalyzer.USER_DEFINED_FUNCTIONS,
+                String.join(".", createFunctionStmt.getNamePath()));
+        udfBuilder.put(functionFullName, createFunctionStmt);
+      } else if (statement.nodeKind() == RESOLVED_QUERY_STMT) {
+        if (!SqlAnalyzer.isEndOfInput(parseResumeLocation)) {
```

Comment:
I can tell line 185 and this line combined together to verify:
1. only one SELECT in a statement list.
2. that SELECT statement should be in the end of list.

But from readability perspective, neither one explicitly tests there are more two SELECT in a list. I am afraid that for people who don't have context to read code here, they could not get the one single SELECT constraint (although it is implied implicitly). 

My suggestion is you only validate `cannot contain more than one SELECT statement` here and leave `Statement list must end in a SELECT statement` to line 185.

Reply:
This is really checking "No additional statements are allowed after a SELECT statement."

Reply:
+1